### PR TITLE
Cudatoolkit v10.1

### PR DESCRIPTION
**Pull request by andersy005**
Checklist
* [x] Used a fork of the feedstock to propose changes
* [x] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by andersy005** [comment]
Addresses #15

**Comment by andersy005** [comment]
@conda-forge-admin, please rerender

File: recipe/scripts/activate.d/cudatoolkit-dev-activate.sh
Comment (pearu) [inline recipe/scripts/activate.d/cudatoolkit-dev-activate.sh:6]: @andersy005 , there is a possibility that downloading cuda installer or when running the installer fails. When such a failure will occur, `$CONDA_PREFIX/pkgs/cuda-toolkit/bin` will not exist and by default, bash will interpret `$CONDA_PREFIX/pkgs/cuda-toolkit/bin/*` as a filename (it does not expand the wildcard). As a result, the `link` value will be `*`, following that the `ln -sf ..` command will actually read `ln -sf $CONDA_PREFIX/pkgs/cuda-toolkit/bin/*  $CONDA_PREFIX/bin/*` . This will fail with a message `ln: target ... is not a directory` (as in issue #13 , I have seen this also when conda packaging omniscidb-cuda).

Long story short, I suggest using `shopt -s nullglob` before the `for` loop and `shopt -u nullglob` after `done`. As a result, the error messages from any failure of running cuda installer should become clearer.
This applies to all `for` loops in this file.

Comment (andersy005) [reply recipe/scripts/activate.d/cudatoolkit-dev-activate.sh:6]: @pearu, thank you for chiming in! I've resolved this. Let me know if there's anything else that needs to be addressed

**Comment by andersy005** [comment]
**Update**

- Everything seems to be working just fine on my local machine (ubuntu 18.04). However, due an old OS (Centos 6) used by the container, there seems to be issues with old glibc version. 


```bash 
++ conda info

     active environment : base
    active env location : /opt/conda
            shell level : 1
       user config file : /home/conda/.condarc
 populated config files : /home/conda/.condarc
          conda version : 4.7.11
    conda-build version : 3.18.9
         python version : 3.7.3.final.0
       virtual packages : 
       base environment : /opt/conda  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /opt/conda/pkgs
                          /home/conda/.conda/pkgs
       envs directories : /opt/conda/envs
                          /home/conda/.conda/envs
               platform : linux-64
             user-agent : conda/4.7.11 requests/2.22.0 CPython/3.7.3 Linux/4.15.0-1052-azure centos/6.10 glibc/2.12
                UID:GID : 1001:1001
             netrc file : None
           offline mode : False
```

```bash
stdout: Running Post installation
downloading https://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda_10.1.243_418.87.00_linux.run to /home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/pkgs/cudatoolkit-dev/cuda_10.1.243_418.87.00_linux.run
Extracting on Linux

stderr: ./cuda-installer: /lib64/libc.so.6: version `GLIBC_2.14' not found (required by ./cuda-installer)
./cuda-installer: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.14' not found (required by ./cuda-installer)
./cuda-installer: /usr/lib64/libstdc++.so.6: version `GLIBCXX_3.4.15' not found (required by ./cuda-installer)
Traceback (most recent call last):
  File "/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/cudatoolkit-dev-post-install.py", line 257, in <module>
    _main()
  File "/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/cudatoolkit-dev-post-install.py", line 253, in _main
    extractor.extract()
  File "/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/cudatoolkit-dev-post-install.py", line 142, in extract
    raise exc
  File "/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/bin/cudatoolkit-dev-post-install.py", line 138, in extract
    subprocess.check_call(cmd, env=dict(DISPLAY=""))
  File "/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/lib/python3.7/subprocess.py", line 347, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '[PosixPath('/home/conda/feedstock_root/build_artifacts/cudatoolkit-dev_1566775645864/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_/pkgs/cudatoolkit-dev/cuda_10.1.243_418.87.00_linux.run'), '--extract=/tmp/tmpd4rou7kq']' returned non-zero exit status 1.

return code: 1
```

File: recipe/meta.yaml
Comment (pearu) [inline recipe/meta.yaml:25]: The `run_test.py` seems to be equivalent to running `nvvm --version`. So, instead of making the test numba (and numpy) dependent, I'd suggest:
```
test:
  commands:
    - nvcc --version
```

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [inline recipe/cudatoolkit-dev-post-install.py]: The try-except block can be removed.

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [inline recipe/cudatoolkit-dev-post-install.py]: I have found `debug_install_path` feature useful to save time in debugging. How about:
```
        local_path = os.path.join(os.path.dirname(os.environ['RECIPE_DIR']),
                                   os.path.basename(dl_path))
        if os.path.isfile(dl_path):
            print("using %s" % (dl_path))
        elif os.path.isfile(local_path):
            print("copying %s to %s" % (local_path, dl_path))
            shutil.copyfile(local_path, dl_path)
        else:
            print("downloading %s to %s" % (dl_url, dl_path))
            self.download(dl_url, dl_path)
```
that will copy the cuda installer from the parent directory of recipe if it exists?

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [inline recipe/cudatoolkit-dev-post-install.py]: Replace with
```
raise NotImplementedError("%s.extract(...)" % (type(self).__name__))
```

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [inline recipe/cudatoolkit-dev-post-install.py]: I think we'll need `--override` option to resolve the following error when extracting:
```
[INFO]: Setting extractPath=/tmp/tmpam64w639
[INFO]: Setting defaultpath=/tmp/tmpam64w639
[INFO]: Checking compiler version...
[INFO]: gcc location: which: no gcc in ((null))

[ERROR]: Missing gcc. gcc is required to continue.

```

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [inline recipe/cudatoolkit-dev-post-install.py]: This try-except block can be removed as well.

File: recipe/scripts/activate.d/cudatoolkit-dev-activate.sh
Comment (pearu) [inline recipe/scripts/activate.d/cudatoolkit-dev-activate.sh]: Remove.

File: recipe/scripts/activate.d/cudatoolkit-dev-activate.sh
Comment (pearu) [inline recipe/scripts/activate.d/cudatoolkit-dev-activate.sh:3]: For a `conda debug` environment, I needed
```
    mkdir -p $CONDA_PREFIX/bin
    mkdir -p $CONDA_PREFIX/lib
    mkdir -p $CONDA_PREFIX/include
```

**Comment by pearu** [comment]
With the `--override` change, the conda-build works in a local Centos 7.5 machine.

File: recipe/cudatoolkit-dev-post-install.py
Comment (pearu) [reply recipe/cudatoolkit-dev-post-install.py]: Although this changeset does not apply when conda-build runs the tests - the cuda installer download is still triggered...

**Comment by pearu** [comment]
@andersy005 could you give me commit rights to https://github.com/andersy005/cudatoolkit-dev-feedstock/ so that I could continue working on this PR in case you are busy?

**Comment by andersy005** [comment]
@pearu, thank you for your patience! I got lost in other stuff at work. Regarding the commit rights, I just sent you an invite: https://github.com/andersy005/cudatoolkit-dev-feedstock/invitations. 

Feel free to ping me if I can be of assistance

**Comment by pearu** [comment]
Thanks @andersy005 , for the access!

Regarding the remaining GLIBC issue, I found this https://devtalk.nvidia.com/default/topic/1050512/cuda-setup-and-installation/error-during-installation-version-glibc_2-14-not-found/

I'll continue looking at the solution that would not require installing glibc..

**Comment by pearu** [comment]
OK, it looks like we are in a dead-end with this PR: cuda installer requires glibc 2.14 but azure CI build machine has glibc 2.12.

On the other hand, the built cudatoolkit-dev conda package should be usable from users machines (that have glibc 2.14 or greater). So, one solution would be to disable testing cudatoolkit-dev installation process on azure machines. This also means that cudatoolkit-dev could not be used as a dependency for packages that are built on azure CI (like omniscidb-cuda). Although, locally, in a user machine the cudatoolkit-dev dependency should work out fine.

Another solution is to downgrade cuda to a version where its installer works with glibc 2.12. v10.0 seems to be ok. This solution would hold back upgrading cuda versions until azure CI machines will have some newer glibc version installed.

@andersy005 @scopatz  Thoughts?

**Comment by andersy005** [comment]
@pearu,

I am in favor of merging this as is even though it will cause other packages that use `cudatoolkit-dev` as a dependency to fail the CI checks for the time being. It would be nice to know whether 

1) there's a plan of upgrading the base docker image's OS (to address the older glibc version issue) anytime soon?

and/or

2) there's another way of addressing the glibc version issue ourselves?

**Comment by pearu** [comment]
Re 1: I don't know, I suspect not soon, otherwise, the CI instances would not be so old..
Re 2: one option is to create glibc conda package or build glibc within the post-installation script. Here are the basic instructions:  https://www.jertype.com/posts/upgrading-glibc/

**Comment by pearu** [comment]
@isuruf thanks for the rhel6 hint!
@andersy005 please review.

**Review by andersy005 — COMMENTED** [review]
Thanks @pearu, @isuruf! Glad to see this brought to completion.
Everything looks good to me. I can merge this soon unless there are any objection.

**Review by xmnlab — APPROVED** [review]
LGTM! thanks everyone for working on that!

**Comment by xmnlab** [comment]
as @andersy005 also approved that, I am merging this PR.

**Pull request merged**
